### ISSUE TITLE
Optional chaining syntax: the ?. operator

### ISSUE THREAD

**Issue by chshersh** [dhall-lang/dhall-lang#1141]
Dhall is amazing when it comes to getting and setting fields of nested records. However, when those fields are wrapped into some types (e.g. `Optional`), then you need to use more verbose standard functions to extract relevant fields.

To address this ergonomic problem, I propose to add the optional chaining operator `?.` for accessing nested optional fields, similar to other languages, e.g. [Kotlin](https://kotlinlang.org/docs/null-safety.html#safe-calls).

> **Note:** I do understand that this doesn't address the more general problem of having other wrapper nested types such as lists, maps, etc. But I believe that `Optional` is a very popular use-case, and improving ergonomics only for this special data type would already be a huge productivity boost.

## Real-life use case

I'm working with Dhall bindings to the OpenShift API, produced by the great `dhall-openapi` tool. And the generated types contain a lot of options. Specifically, one example from our code base:

* `target.dhall`
    ```dhall
    { metadata : API.ObjectMeta.Type
    , spec : Optional API.PodTemplateSpec.Type
    }
    ```
* `PodTemplateSpec.dhall`
    ```dhall
    { metadata : Optional ./io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta.dhall
    , spec : Optional ./io.k8s.api.core.v1.PodSpec.dhall
    }
    ```
* `ObjectMeta.dhall`
    ```dhall
    { labels : Optional (List { mapKey : Text, mapValue : Text })
    , ...
    }
    ```

I want to extract `labels` from the `spec` field of the `target.dhall` type. My current code looks like this:

```dhall
let selector =
      Prelude.Optional.concatMap
        API.PodTemplateSpec.Type
        (List { mapKey : Text, mapValue : Text })
        ( λ(podSpec : API.PodTemplateSpec.Type) →
            Prelude.Optional.concatMap
              API.ObjectMeta.Type
              (List { mapKey : Text, mapValue : Text })
              (λ(metadata : API.ObjectMeta.Type) → metadata.labels)
              podSpec.metadata
        )
        params.target.spec
```

With the optional `?.` chaining operator it becomes:

```dhall
let selector = params.target.spec?.metadata?.labels
```

## Proposed change

Introduce the optional chaining `?.` operator for getting fields of nested `Optional` fields inside records.

Semantics:

* Priority is the same as for ordinary `.`, so they can be mixed together.
    * The parenthesis in my example above: `(((params.target).spec)?.metadata)?.labels`
* When at least one of the fields in the chain is `None`, the result will be `None`. Otherwise, return `Some` with the value of the last field specified

The `?.` operator can be a syntax sugar for `Optional.concatMap`, but that's implementation details.

**Comment by TristanCacqueray** [comment]
That seems related to https://github.com/dhall-lang/dhall-lang/issues/979 and https://github.com/dhall-lang/dhall-lang/issues/924 . There was also some earlier discussion about `?.` in https://github.com/dhall-lang/dhall-lang/pull/923 . It seems like such operator would be very helpful when working with kubernetes resources.

**Comment by Gabriel439** [comment]
Yeah, I support something like this and agree that it overlaps with #924

**Comment by chshersh** [comment]
@TristanCacqueray @Gabriel439 Thanks for getting back to this so quickly! And thanks for the links to the existing discussions. It's nice to see that similar ideas already appeared before.

**Comment by ari-becker** [comment]
Something like this is practically a necessity for working with types like those in `dhall-kubernetes`. The example @chshersh provided is really only the tip of the iceberg.

Unfortunately my understanding is that what holds this up is that it practically requires type inference, i.e. that it can't just be provided as syntactic sugar because (in the example provided) it can't infer either `API.PodTemplateSpec.Type` or `API.ObjectMeta.Type` from just `params.target.spec?.metadata?.labels`. Maybe there's something the implementations can do that doesn't require type inference, or maybe type inference is farther along than I'm familiar with currently. Would love to be wrong.

**Comment by TristanCacqueray** [comment]
Without type inference, could the type of `a.b?.c?.d` simply be `Optional DType`, where if any of the `?.` operands is None, then the result would be `None DType` ?

For example:
```haskell
let Pod = { spec : Optional { metadata : Optional { value : Text } } }

let emptyPod : Pod
    = { spec = None { metadata : Optional { value : Text } } }

let emptyMetadata : Pod
    = { spec = Some { metadata = None { value : Text } } }

let podWithMeta : Pod
    = { spec = Some { metadata = Some { value = "value" } } }

in  [
         emptyPod.spec?.metadata?.value -- None Text                                                                                                                                                                                                                                                                                                                                                                                         
  , emptyMetadata.spec?.metadata?.value -- None Text                                                                                                                                                                                                                                                                                                                                                                                         
  ,   podWithMeta.spec?.metadata?.value -- Some "value"                                                                                                                                                                                                                                                                                                                                                                                      
] : List (Optional Text)
```

**Comment by Gabriel439** [comment]
I put up a funding proposal for standardizing this here: https://discourse.dhall-lang.org/t/expense-proposal-add-with-support-for-modifying-optional-values/520

**Comment by darichey** [comment]
I've implemented the functionality for `with` in #1254!
